native/dlasq3.go

```diff
@@ -0,0 +1,153 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "math"
+
+// Dlasq3 checks for deflation, computes a shift (tau) and calls dqds.
+// In case of failure it changes shifts, and tries again until output
+// is positive.
+func (impl Implementation) Dlasq3(i0, n0 int, z []float64, pp int, dmin, sigma, desig, qmax float64, nFail, iter, nDiv int, ttype int, dmin1, dmin2, dn, dn1, dn2, g, tau float64) (
+	i0Out, n0Out, ppOut int, dminOut, sigmaOut, desigOut, qmaxOut float64, nFailOut, iterOut, nDivOut, ttypeOut int, dmin1Out, dmin2Out, dnOut, dn1Out, dn2Out, gOut, tauOut float64) {
+	// Note: The documentation says dmin is an output only, but if that is so it is
+	// used before it is set.
+	const cbias = 1.5
+
+	n0in := n0
+	eps := dlamchP
+	tol := eps * 100
+	tol2 := tol * tol
+	var nn int
+	var t float64
+Ten:
+	if n0 < i0 {
+		return i0, n0, pp, dmin, sigma, desig, qmax, nFail, iter, nDiv, ttype, dmin1, dmin2, dn, dn1, dn2, g, tau
+	}
+	if n0 == i0 {
+		goto Twenty
+	}
+	nn = 4*(n0+1) + pp - 1
+	if n0 == i0+1 {
+		goto Fourty
+	}
+	if z[nn-5] > tol2*(sigma+z[nn-3]) && z[nn-2*pp-4] > tol2*z[nn-7] {
+		goto Thirty
+	}
+Twenty:
+	z[4*(n0+1)-4] = z[4*(n0+1)+pp-4] + sigma
+	n0--
+	goto Ten
+Thirty:
+	if z[nn-9] > tol2*sigma && z[nn-2*pp-8] > tol2*z[nn-11] {
+		goto Fifty
```

Comment:
Apparently unused goto lables are also a compile error. I'm not sad I haven't encountered this before.